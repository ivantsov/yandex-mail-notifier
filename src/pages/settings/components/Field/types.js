export default {
  checkbox: 'checkbox',
  select: 'select',
  link: 'link',
};
