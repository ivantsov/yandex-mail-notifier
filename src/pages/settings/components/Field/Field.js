import React, {PropTypes} from 'react';
import types from './types';
import Checkbox from './Items/Checkbox';
import Select from './Items/Select';

import styles from './Field.less';

const Field = (props) => {
    const {
        type,
        name,
        label,
        description,
    } = props;

    const Component = type === types.checkbox ? Checkbox : Select;

    return (
        <tr>
            <td className={styles.label}>
                <label htmlFor={name}>{label}</label>
                {description && <div className={styles.description}>{description}</div>}
            </td>
            <td>
                <Component {...props}/>
            </td>
        </tr>
    );
};

Field.propTypes = {
    type: PropTypes.oneOf(Object.keys(types)).isRequired,
    name: PropTypes.string.isRequired,
    label: PropTypes.string.isRequired,
    description: PropTypes.string.isRequired,
};

export default Field;
