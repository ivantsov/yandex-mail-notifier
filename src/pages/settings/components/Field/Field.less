.label {
    padding: 8px 16px 8px 0;
}

.description {
    color: #888;
}
