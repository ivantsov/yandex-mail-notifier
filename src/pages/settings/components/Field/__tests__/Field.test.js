import React from 'react';
import renderer from 'react-test-renderer';
import Field from '../Field';

jest.mock('../Items/Checkbox', () => (props) => <div {...props}>Checkbox</div>);
jest.mock('../Items/Select', () => (props) => <div {...props}>Select</div>);
jest.mock('../Items/Link', () => (props) => <div {...props}>Link</div>);

function testCase(props) {
    const component = renderer.create(
        <Field
            name="field"
            label="field label"
            description="field description"
            {...props}
        />,
    );

    const tree = component.toJSON();
    expect(tree).toMatchSnapshot();
}

describe('settings/Field', () => {
    it('checkbox', () => {
        testCase({
            type: 'checkbox',
            value: false,
            onChange: jest.fn(),
        });
    });

    it('select', () => {
        testCase({
            type: 'select',
            value: 1,
            onChange: jest.fn(),
        });
    });

    it('link', () => {
        testCase({
            type: 'link',
            url: 'https://smth.com',
            text: 'site url',
        });
    });
});
