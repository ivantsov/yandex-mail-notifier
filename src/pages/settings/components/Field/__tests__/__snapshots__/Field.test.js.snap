// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`settings/Field checkbox 1`] = `
<tr>
  <td
    className="label"
  >
    <label
      htmlFor="field"
    >
      field label
    </label>
    <div
      className="description"
    >
      field description
    </div>
  </td>
  <td>
    <div
      description="field description"
      label="field label"
      name="field"
      onChange={[Function]}
      type="checkbox"
      value={false}
    >
      Checkbox
    </div>
  </td>
</tr>
`;

exports[`settings/Field link 1`] = `
<tr>
  <td
    className="label"
  >
    <label
      htmlFor="field"
    >
      field label
    </label>
    <div
      className="description"
    >
      field description
    </div>
  </td>
  <td>
    <div
      description="field description"
      label="field label"
      name="field"
      text="site url"
      type="link"
      url="https://smth.com"
    >
      Link
    </div>
  </td>
</tr>
`;

exports[`settings/Field select 1`] = `
<tr>
  <td
    className="label"
  >
    <label
      htmlFor="field"
    >
      field label
    </label>
    <div
      className="description"
    >
      field description
    </div>
  </td>
  <td>
    <div
      description="field description"
      label="field label"
      name="field"
      onChange={[Function]}
      type="select"
      value={1}
    >
      Select
    </div>
  </td>
</tr>
`;
